### PR TITLE
Add insert support to iptables.

### LINKED ISSUE #2167: iptables module: add possibility to INSERT rules (right now is only APPEND)
##### ISSUE TYPE

- Feature Idea
##### COMPONENT NAME

iptables
##### ANSIBLE VERSION

```
ansible 2.0.1.0
```
##### OS / ENVIRONMENT

N/A

### PR DESCRIPTION
**Pull request by dvigueras**
This PR adds support to be able to insert new rules in iptables (iptables -I).

I think @LinusU should review this PR as he is the author of the iptables module.

Fixes #2167

### CODE REVIEW

**Comment by bcoca** [comment]
what is the use case here? 
would the same behavior not be achieved with ensuring that the rule is absent and then present?

File: system/iptables.py [hunk 1]
```diff
@@ -220,7 +234,6 @@ def append_param(rule, param, flag, is_list):
         if param is not None:
             rule.extend([flag, param])
 
-
```
Comment (LinusU) [inline system/iptables.py:300]: Why remove this line?

File: system/iptables.py [hunk 2]
```diff
@@ -296,12 +315,13 @@ def main():
         state=module.params['state'],
     )
     ip_version = module.params['ip_version']
+    insert     = module.params['insert']
```
Comment (LinusU) [inline system/iptables.py]: PEP 8 don't like the multiple spaces, personally I would have removed them and also put this line above the previous one...

**Comment by dvigueras** [comment]
It isn't the same behavior.
Suppose the following scenario:

Iptables INPUT rule list:
- Rule A
- Rule B

If we ensure rule A is absent, and then present, the result would be:
- Rule B
- Rule A

But what we can achieve with this new insert feature is to be able to insert the rule A at the top of the chain, so with insert=yes the result would be:
- Rule A
- Rule B.

**Comment by LinusU** [comment]
I had a few nits but the code looks good. I have two questions though.

Why is it named `insert`?

What is the use case?

**Comment by bcoca** [comment]
so basically this is insert=top/bottom with bottom being the default?

**Comment by LinusU** [comment]
Ah, now I understand why :)

Could we have a parameter that says wether to insert it at the top or the bottom instead?

File: system/iptables.py (hunk 1 above)
Comment (dvigueras) [inline system/iptables.py:300]: It was a typo, fixed.

File: system/iptables.py (hunk 2 above)
Comment (dvigueras) [inline system/iptables.py]: Reordered vars assignment and removed extra spaces.

**Comment by dvigueras** [comment]
I think there are two options:
- Option A) As it is defined now. There is a boolean insert parameter which controls how iptables should be called. insert=yes means iptables will be called with the -I option, and insert=no (default value) means iptables will be called with the -A option.
- Option B) As @bcoca says. There is a string insert parameter which controls how iptables should be called, the possible values are top and bottom. insert=top means iptables will be called with the -I option, and insert=bottom (default value) means iptables will be called with the -A option.

I think both options are valid. Opinions?

**Comment by gregdek** [comment]
@LinusU thanks for your input so far. Please be sure to review this PR to make sure it adheres to the following guidelines:

http://docs.ansible.com/developing_modules.html#module-checklist

If it passes these guidelines, and if you believe it’s a good PR otherwise, please add a comment with "shipit" in the text, and we will flag it for inclusion.

If it fails these guidelines, please add a comment with "needs_revision" in the text, along with the needed changes, and we will flag it for revision.

**Comment by LinusU** [comment]
Hmm, one more thing that I don't like. `insert` moves the rule to the top should it already exist, `append` on the other hand doesn't move the rule to the bottom if it already exists. That feels very unintuitive to me.

What I propose:

One parameter named `action`, with the possible values: `append`, `insert`.

Possibly an extra parameter that controls wether or not the rule should be relocated if it isn't at the desired location. If we implement this, it needs to output correct change information, i.e. if the rule already was in the correct position, it should not be marked as changed.

Do we need the second feature? Could we split it into two pull requests with two different discussions?

I think that the first one is a no-brainer and it's very easy to implement. The other one I don't know the demand for and it's a bit tougher to implement.

**Comment by bcoca** [comment]
The insert=top/bottom can be made to only be used when state='present' and the rule is not present already. The other possibility is adding the rule at a specific line number, which should 'move' existing rules.

**Comment by LinusU** [comment]
I agree. Specific line number though came up when I first wrote this module and the problem is that this is the global list on the entire machine. It's not often that you know what number to put in and when you do, it's quite fragile since that number is prone to changing.

**Comment by dvigueras** [comment]
Renamed parameter `insert` to `action` as @LinusU commented.
Possible `action` values are `append` or `insert`. Default value is `append` in order to be backward compatible.

**Comment by dvigueras** [comment]
Hi, It's been 9 days since my last comment. Any news about this?

**Comment by bcoca** [comment]
sorry, we are a very busy project so sometimes things take a while. Also we are concentrating on releasing 2.0 which means bugs take priority and features might be delayed for 2.1

**Comment by dvigueras** [comment]
Ok no problem, I can wait :) I was just waiting @LinusU's opinion.

**Comment by LinusU** [comment]
There is still a few problems to be resolved here. The documentation still says `insert` instead of `action`, and the largest problem:

It's very unintuitive that insert _always changes_, even if the rule was previously on top. In my opinion, the rules should _only change_ when the rules _was absent_.

**Comment by ctlajoie** [comment]
Will we be able to insert a rule at a specific index? From what I've read here so far it looks like the only options are top/bottom.

**Comment by dvigueras** [comment]
@LinusU I agree with you, insert shouldn't change the state always. It should check if the rule is present at the top, but I don't know how to implement it in a clean way (i.e. not parsing `iptables -L` command output). 

With `iptables -C` you can check if a rule is present (as you have already done in the module), but with this command it isn't possible to find out which position is the rule at. Any ideas about how to implement it in a clean way?

@ctlajoie that should be part of another PR and would be pretty easy to add.

**Comment by LinusU** [comment]
I think that checking if the rule is present is good. It would be unexpected if `insert` always move is to the top, even if it's present, and `append` adds the rule at the bottom, but doesn't move it if it's already present.

The behaviour should be the same for both commands, and I think that only inserting it if it isn't present is best. Consider what would happen if you are inserting two rules, they would always change since the first one would be moved to under the second one, and then the second one would be moved to be last again.

**Comment by dvigueras** [comment]
@LinusU done, and I also fixed the documentation mistake. 

Now if you insert an already present rule it won't change the chain.

**Comment by LinusU** [comment]
@dvigueras Looks very good :+1:

**Comment by LinusU** [comment]
Hope I'm doing this right now:

'shipit'

**Comment by gregdek** [comment]
Thanks again to @dvigueras for this PR, and thanks @LinusU for reviewing. Marking for inclusion.

**Comment by LinusU** [comment]
@dvigueras Why did you close this? Could you reopen, we would love to get this in 👍

**Comment by dvigueras** [comment]
Maybe I deleted the branch accidentally, sorry. I'll reopen the PR but I think it will require a rebase.

**Comment by LinusU** [comment]
Thanks!

'shipit'

**Pull request merged**